src/com/amazon/ion/impl/lite/IonValueLite.java

```diff
@@ -1002,19 +985,59 @@ final void writeTo(IonWriter writer, SymbolTableProvider symbolTableProvider)
             writer.setFieldNameSymbol(tok);
         }
 
-        SymbolToken[] annotations = getTypeAnnotationSymbols(symbolTableProvider);
+        SymbolToken[] annotations = value.getTypeAnnotationSymbols(symbolTableProvider);
         writer.setTypeAnnotationSymbols(annotations);
+    }
 
-        try
-        {
-            writeBodyTo(writer, symbolTableProvider);
-        }
-        catch (IOException e)
-        {
+    private void writeToIterative(IonWriter writer, SymbolTableProvider symbolTableProvider) throws IOException {
+        Deque<Iterator<IonValue>> iteratorStack = null;
+        Iterator<IonValue> currentIterator = null;
+        IonValueLite value = this;
+        do {
+            writeFieldNameAndAnnotations(writer, value, symbolTableProvider);
+            IonType type = value.getType();
+            if (value.isNullValue()) {
+                writer.writeNull(type);
+            } else if (IonType.isContainer(type)) {
+                if (currentIterator != null) {
+                    if (iteratorStack == null) {
+                        iteratorStack = new ArrayDeque<>(CONTAINER_STACK_INITIAL_CAPACITY);
+                    }
+                    iteratorStack.add(currentIterator);
+                }
+                currentIterator = ((IonContainer) value).iterator();
+                writer.stepIn(type);
+            } else {
+                value.writeBodyTo(writer, symbolTableProvider);
+            }
+            do {
+                if (currentIterator == null) {
+                    return;
+                }
+                value = currentIterator.hasNext() ? (IonValueLite) currentIterator.next() : null;
+                if (value == null) {
+                    writer.stepOut();
+                    currentIterator = iteratorStack == null ? null : iteratorStack.pollLast();
+                }
+            } while (value == null);
+        } while (true);
```

Comment:
When I know I have a loop that needs to execute unconditionally at least once and I'm not sure exactly what the continuation condition will be, I often start it as a `do / while` rather than a `while(true)`. I don't consider one to be more or less visually appealing than the other.